File: python/2d-dynamic-programming/regular-expression-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        cache = {}

        def dfs(i, j):
            print(i, j)
            if i >= len(s) and j >= len(p):
                return True
            if j >= len(p):
                return False
            if (i, j) in cache:
                return cache[(i, j)]

            match = i < len(s) and (s[i] == p[j] or p[j] == ".")
            if j < len(p) - 1 and p[j + 1] == "*":
                cache[(i, j)] = (match and (dfs(i + 1, j) or dfs(i, j + 2))) or dfs(i, j + 2)
            else:
                cache[(i, j)] = match and dfs(i + 1, j + 1)

            return cache[(i, j)]

        return dfs(0, 0)

File: c++/2d-dynamic-programming/regular-expression-matching.cc
class Solution {
public:
    bool isMatch(string s, string p) {
        std::unordered_map<int, std::unordered_map<int, bool>> cache;
        return match(0, 0, s, p, cache);
    }

private:
    bool match(int i, int j, std::string& s, std::string& p, std::unordered_map<int, std::unordered_map<int, bool>>& cache) {
        if (i == s.size() && j == p.size()) {
            return true;
        }

        if (j == p.size()) {
            return false;
        }

        if (j < p.size() - 1 && p[j + 1] == '*') {
            if (i < s.size() && (p[j] == s[i] || p[j] == '.')) {
                cache[i][j] = match(i + 1, j, s, p, cache) || match(i, j + 2, s, p, cache);
            } else {
                cache[i][j] = match(i, j + 2, s, p, cache);
            }
        } else {
            if (i < s.size() && (p[j] == s[i] || p[j] == '.')) {
                cache[i][j] = match(i + 1, j + 1, s, p, cache);
            } else {
                cache[i][j] = false;
            }
        }

        return cache[i][j];
    } 
};
